internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -269,15 +289,19 @@ func (u *Upgrader) sourceURI(retrievedURI string) string {
 	return u.settings.SourceURI
 }
 
-func rollbackInstall(ctx context.Context, log *logger.Logger, hash string) {
-	os.RemoveAll(filepath.Join(paths.Data(), fmt.Sprintf("%s-%s", agentName, hash)))
+func rollbackInstall(ctx context.Context, log *logger.Logger, versionedHome string) {
+	err := os.RemoveAll(filepath.Join(paths.Top(), versionedHome))
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		// TODO should this be a warning or an error ?
+		log.Warnw("error rolling back install", "error.message", err)
```

Comment:
I think warning is fine.  It's better than before this PR when we weren't even logging anything if `os.RemoveAll` returned an error.

Also, from my understanding, this function is being called when one of the upgrade steps fails and we want to clean up the target agent's files. The current (pre-upgrade) Agent is still running fine at this point and the callers of this function will return an error about the upgrade step that failed anyway.  So I think just a warning here is fine.

Reply:
I'd vote error so we can also see it happening, but it also isn't fatal, so technically warning is fine.

I think the error message should note that the error was specifically from `os.RemoveAll`. That doesn't mean that changing the symlink failed, that is the fatal part of rollbackInstall. We could change this to "failed to remove %directory contents" instead.

If this does fail though, does the agent directory that was left behind just stick around forever? If it does that is something for us to follow up on separately.

Reply:
@cmacknz
Any leftover directories  other than the one we want to keep gets removed during watcher Cleanup() or Rollback() calls: this is the reason why at this stage watcher breaks the agent install (it still removes anything that is not `elastic-agent-<hash we want to keep>`).

About the error/warning and the message, it has been changed in PR #4178 https://github.com/elastic/elastic-agent/pull/4178/files/6fd94ced07902a2a9baf79ce3e695251b8f63468..b60b518fcaa18ccac4e8801ff1f07090adc0690d#diff-42a05457d0720cf1e61083ad5d04a0b2956e6cfa0e0d018f2f815327cb0f1d2dR377-R383
where the logs are both at warn level and the message explicitly say what went wrong